pkg/test/runner/config.go

```diff
@@ -23,13 +23,39 @@ import (
 	"gopkg.in/yaml.v3"
 )
 
+// EvalTestCaseConfig contains the config only for imperative
+// function run
+type EvalTestCaseConfig struct {
+	// ExecPath is a path to the executable file that will be run as function
+	// Mutually exclusive with Image
+	ExecPath string `json:"execPath,omitempty" yaml:"execPath,omitempty"`
+	// Image is the image name for the function
+	Image string `json:"image,omitempty" yaml:"image,omitempty"`
+	// Args are the arguments that will be passed into function.
+	// Args will be passed as 'key=value' format after the '--' in command.
+	Args map[string]string `json:"args,omitempty" yaml:"args,omitempty"`
+	// Network indicates is network accessible from the function container. Default: false
+	Network bool `json:"network,omitempty" yaml:"network,omitempty"`
+	// FnConfig is the path to the function config file.
+	FnConfig string `json:"fnConfig,omitempty" yaml:"fnConfig,omitempty"`
+}
+
 // TestCaseConfig contains the config information for the test case
 type TestCaseConfig struct {
-	ExitCode int  `json:"exitCode,omitempty" yaml:"exitCode,omitempty"`
-	Network  bool `json:"network,omitempty" yaml:"network,omitempty"`
-	RunCount int  `json:"runCount,omitempty" yaml:"runCount,omitempty"`
-	Skip     bool `json:"skip,omitempty" yaml:"skip,omitempty"`
-	Debug    bool `json:"debug,omitempty" yaml:"debug,omitempty"`
+	// ExitCode is the expected exit code from the kpt commands. Default: 0
+	ExitCode int `json:"exitCode,omitempty" yaml:"exitCode,omitempty"`
+	// RunCount is how many times the test will run. Each iteration will be
+	// immediately run after previous one. Default: 1
```

Comment:
Shouldn't the default be 2 to test idempotency by default?

Reply:
@natasha41575 changed it to 1, https://github.com/GoogleContainerTools/kpt/pull/1380#discussion_r570591557. But yes we should set it to 2 by default.

Reply:
I will leave this value to 1 and if we really want to change it to 2 we can have another PR.

Reply:
Default (expected) behavior for a function is to be idempotent. This should be reflected in the default values here. In fact, we probably want to have an explicit `idempotent` field with default to true (Unless there's a use case for runCount besides testing for idempotency). Not having idempotency should be an exception and needs explicit/obvious flag. You can follow up in a separate PR and change existing test packages.

Reply:
+1 on naming it `idempotent`.

My .2 cents:
As you said, `idempotency` is the behavior of the function not `fn eval` or `fn render` functionality. So having this default makes sense for function-catalog tests, not for `fn render` or `fn eval` tests. Running `fn render` or `fn eval` tests twice for every tests is going to double the time during dev cycles especially you can not target specific test using `go test -run=TestXXX`.

Reply:
Idempotency is the desired property of using kpt CLI and sinking in-place, so you cannot ask all function developer to go out of their way to test for it. That's an unstable system.

Reply:
Something I am thinking about from your comment: if it is part of the CLI, then why isn't it built in the CLI itself (CLI should be running the pipeline twice to ensure idempotency ?). Tests should be testing the functionality built-in the CLI and introducing the performance overhead (running twice) for something that's not part of the CLI did't feel like the right approach.

I think it's a debatable topic, would follow up offline.